File: web_demo/static/cat-blink-v2.js
/**
 * Cat Blink v2 — fountain-coded optical blink transport codec.
 *
 * Shared reference implementation for the wire format specified in
 * docs/CAT_BLINK_V2.md. Reuses the existing fountain droplet layout
 * (fountain-codes.js) so the blink path and QR path share one FEC layer.
 *
 * Pure functions over payload bytes ⇄ whitened bitstring. The physical blink
 * packing (eye states, preamble, loop timing) lives in the emitter/sampler;
 * this module is the byte/bit/frame/FEC layer and is unit-testable without a
 * camera.
 *
 * Works in the browser (loaded after fountain-codes.js) and in Node
 * (require('./fountain-codes.js')).
 */
(function (root, factory) {
  let Fountain;
  if (typeof module !== 'undefined' && module.exports) {
    Fountain = require('./fountain-codes.js');
    module.exports = factory(Fountain);
  } else {
    // Browser: fountain-codes.js defines these globals.
    Fountain = {
      FountainEncoder: root.FountainEncoder,
      FountainDecoder: root.FountainDecoder,
      Droplet: root.Droplet,
    };
    root.CatBlinkV2 = factory(Fountain);
  }
})(typeof self !== 'undefined' ? self : this, function (Fountain) {
  'use strict';

  const { FountainEncoder, FountainDecoder, Droplet } = Fountain;

  const SYNC = [0xb1, 0x17, 0x5e];
  const VERSION = 0x02;
  const HEADER_LEN = SYNC.length + 1 + 2 + 2 + 4; // sync+ver+k+bs+origlen = 12
  const CRC_LEN = 2;
  const WHITEN_SEED = 0x4d454f57; // "MEOW"

  // ── CRC-16/CCITT-FALSE ──────────────────────────────────────────────────
  function crc16(bytes, start, end) {
    let crc = 0xffff;
    for (let i = start; i < end; i++) {
      crc ^= bytes[i] << 8;
      for (let b = 0; b < 8; b++) {
        crc = crc & 0x8000 ? ((crc << 1) ^ 0x1021) & 0xffff : (crc << 1) & 0xffff;
      }
    }
    return crc & 0xffff;
  }

  // ── MEOW-LCG whitening (self-inverse), reseeded per call ─────────────────
  // Math.imul for the correct low-32-bit product: `seed * 1103515245`
  // overflows Number's 53-bit safe range, so a plain multiply desyncs from
  // the Python/mobile references. Matches catWhitening.ts and cat_video.py.
  function whitenBits(bitStr) {
    let seed = WHITEN_SEED;
    let out = '';
    for (let i = 0; i < bitStr.length; i++) {
      const lo = Math.imul(seed, 1103515245) >>> 0;
      seed = (lo + 12345) & 0x7fffffff;
      const mask = (seed >> 16) & 1;
      out += String((bitStr.charCodeAt(i) - 48) ^ mask);
    }
    return out;
  }
  const dewhitenBits = whitenBits; // self-inverse

  // ── byte ⇄ MSB-first bit helpers ─────────────────────────────────────────
  function bytesToBits(bytes) {
    let s = '';
    for (let i = 0; i < bytes.length; i++) {
      s += bytes[i].toString(2).padStart(8, '0');
    }
    return s;
  }
  function bitsToBytes(bitStr) {
    const n = bitStr.length >> 3;
    const out = new Uint8Array(n);
    for (let i = 0; i < n; i++) {
      out[i] = parseInt(bitStr.substr(i * 8, 8), 2);
    }
    return out;
  }

  // ── Frame one droplet ────────────────────────────────────────────────────
  function frameDroplet(dropletBytes, kBlocks, blockSize, originalLength) {
    const body = new Uint8Array(HEADER_LEN - SYNC.length + dropletBytes.length);
    const dv = new DataView(body.buffer);
    let o = 0;
    body[o++] = VERSION;
    dv.setUint16(o, kBlocks, false); o += 2;
    dv.setUint16(o, blockSize, false); o += 2;
    dv.setUint32(o, originalLength, false); o += 4;
    body.set(dropletBytes, o);

    const frame = new Uint8Array(SYNC.length + body.length + CRC_LEN);
    frame.set(SYNC, 0);
    frame.set(body, SYNC.length);
    const crc = crc16(frame, SYNC.length, SYNC.length + body.length);
    frame[frame.length - 2] = (crc >> 8) & 0xff;
    frame[frame.length - 1] = crc & 0xff;
    return frame;
  }

  /**
   * Encode a payload into `frameCount` whitened blink frames (bitstrings).
   * Returns { frames: string[], kBlocks, blockSize, originalLength }.
   */
  function encode(payloadBytes, opts = {}) {
    const blockSize = opts.blockSize || 32;
    const redundancy = opts.redundancy || 3.0;
    const minFrames = opts.minFrames || 8;
    const kBlocks = Math.max(1, Math.ceil(payloadBytes.length / blockSize));
    const frameCount = Math.max(minFrames, Math.ceil(kBlocks * redundancy));

    const enc = new FountainEncoder(payloadBytes, kBlocks, blockSize);
    const frames = [];
    for (let i = 0; i < frameCount; i++) {
      const droplet = enc.generateDroplet(i);
      const frameBytes = frameDroplet(droplet.pack(), kBlocks, blockSize, payloadBytes.length);
      frames.push(whitenBits(bytesToBits(frameBytes)));
    }
    return { frames, kBlocks, blockSize, originalLength: payloadBytes.length, frameCount };
  }

  /**
   * Parse a single whitened frame bitstring back to its droplet + params.
   * Returns null if SYNC/version/CRC fail. Length-agnostic: reads exactly the
   * droplet the header describes.
   */
  function decodeFrame(whitenedBits) {
    if (whitenedBits.length < (HEADER_LEN + CRC_LEN) * 8) return null;
    const bits = dewhitenBits(whitenedBits);
    const bytes = bitsToBytes(bits);
    if (bytes[0] !== SYNC[0] || bytes[1] !== SYNC[1] || bytes[2] !== SYNC[2]) return null;
    const dv = new DataView(bytes.buffer);
    let o = SYNC.length;
    const version = bytes[o]; o += 1;
    if (version !== VERSION) return null;
    const kBlocks = dv.getUint16(o, false); o += 2;
    const blockSize = dv.getUint16(o, false); o += 2;
    const originalLength = dv.getUint32(o, false); o += 4;
    // droplet: seed(4)+count(2)+indices(count*2)+data(blockSize)
    const dropStart = o;
    if (bytes.length < dropStart + 6) return null;
    const count = dv.getUint16(dropStart + 4, false);
    const dropLen = 4 + 2 + count * 2 + blockSize;
    const crcStart = dropStart + dropLen;
    if (bytes.length < crcStart + CRC_LEN) return null;
    const got = (bytes[crcStart] << 8) | bytes[crcStart + 1];
    const want = crc16(bytes, SYNC.length, crcStart);
    if (got !== want) return null;
    const dropletBytes = bytes.subarray(dropStart, dropStart + dropLen);
    return { version, kBlocks, blockSize, originalLength, dropletBytes, count };
  }

  /**
   * Decode a set of whitened frame bitstrings into the original payload.
   * Frames may be in any order and a lossy subset — the fountain layer
   * reconstructs from any sufficient set. Returns Uint8Array or null.
   */
  function decode(whitenedFrames) {
    let decoder = null;
    let originalLength = null;
    for (const wf of whitenedFrames) {
      const f = decodeFrame(wf);
      if (!f) continue; // dropped/corrupt frame — fountain tolerates it
      if (!decoder) {
        decoder = new FountainDecoder(f.kBlocks, f.blockSize, f.originalLength);
        originalLength = f.originalLength;
      }
      const droplet = Droplet.unpack(f.dropletBytes, f.blockSize);
      decoder.addDroplet(droplet);
      if (decoder.isComplete()) break;
    }
    if (!decoder || !decoder.isComplete()) return null;
    return decoder.getData(originalLength);
  }

  return {
    SYNC, VERSION, HEADER_LEN, CRC_LEN,
    crc16, whitenBits, dewhitenBits, bytesToBits, bitsToBytes,
    frameDroplet, encode, decodeFrame, decode,
  };
});

File: meow_decoder/cat_blink_v2.py
"""Cat Blink v2 — fountain-coded optical blink transport codec (Python port).

Reference implementation of the wire format in docs/CAT_BLINK_V2.md, matching
web_demo/static/cat-blink-v2.js bit-for-bit. Reuses the existing fountain
droplet layout via meow_decoder.fountain, so the blink path and QR path share
one FEC layer.

Pure functions over payload bytes <-> whitened bitstrings. The physical blink
packing (eye states, preamble, loop timing) lives in the emitter/sampler; this
module is the byte/bit/frame/FEC layer and is unit-testable without a camera.
"""

from __future__ import annotations

import struct
from typing import List, Optional

from meow_decoder.fountain import (
    FountainDecoder,
    FountainEncoder,
    pack_droplet,
    unpack_droplet,
)

SYNC = bytes((0xB1, 0x17, 0x5E))
VERSION = 0x02
_HEADER_LEN = len(SYNC) + 1 + 2 + 2 + 4  # sync+ver+k+bs+origlen = 12
_CRC_LEN = 2
_WHITEN_SEED = 0x4D454F57  # "MEOW"


def crc16(data: bytes) -> int:
    """CRC-16/CCITT-FALSE: poly 0x1021, init 0xFFFF, no reflection/final-xor."""
    crc = 0xFFFF
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            crc = ((crc << 1) ^ 0x1021) & 0xFFFF if crc & 0x8000 else (crc << 1) & 0xFFFF
    return crc & 0xFFFF


def whiten_bits(bit_str: str) -> str:
    """XOR a bitstring with the MEOW-LCG keystream (self-inverse, reseeded)."""
    seed = _WHITEN_SEED
    out = []
    for ch in bit_str:
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        mask = (seed >> 16) & 1
        out.append(str((ord(ch) - 48) ^ mask))
    return "".join(out)


dewhiten_bits = whiten_bits  # self-inverse


def bytes_to_bits(data: bytes) -> str:
    return "".join(f"{b:08b}" for b in data)


def bits_to_bytes(bit_str: str) -> bytes:
    n = len(bit_str) // 8
    return bytes(int(bit_str[i * 8 : i * 8 + 8], 2) for i in range(n))


def frame_droplet(
    droplet_bytes: bytes, k_blocks: int, block_size: int, original_length: int
) -> bytes:
    body = (
        bytes((VERSION,))
        + struct.pack(">H", k_blocks)
        + struct.pack(">H", block_size)
        + struct.pack(">I", original_length)
        + droplet_bytes
    )
    crc = crc16(body)
    return SYNC + body + struct.pack(">H", crc)


def encode(
    payload: bytes,
    block_size: int = 32,
    redundancy: float = 3.0,
    min_frames: int = 8,
) -> List[str]:
    """Encode a payload into whitened blink-frame bitstrings."""
    import math

    k_blocks = max(1, math.ceil(len(payload) / block_size))
    frame_count = max(min_frames, math.ceil(k_blocks * redundancy))
    enc = FountainEncoder(payload, k_blocks, block_size)
    frames = []
    for i in range(frame_count):
        droplet = enc.droplet(i)
        frame = frame_droplet(pack_droplet(droplet), k_blocks, block_size, len(payload))
        frames.append(whiten_bits(bytes_to_bits(frame)))
    return frames


def decode_frame(whitened_bits: str) -> Optional[dict]:
    """Parse one whitened frame; return None on SYNC/version/CRC failure."""
    if len(whitened_bits) < (_HEADER_LEN + _CRC_LEN) * 8:
        return None
    data = bits_to_bytes(dewhiten_bits(whitened_bits))
    if data[: len(SYNC)] != SYNC:
        return None
    o = len(SYNC)
    if data[o] != VERSION:
        return None
    o += 1
    k_blocks = struct.unpack_from(">H", data, o)[0]
    o += 2
    block_size = struct.unpack_from(">H", data, o)[0]
    o += 2
    original_length = struct.unpack_from(">I", data, o)[0]
    o += 4
    drop_start = o
    if len(data) < drop_start + 6:
        return None
    count = struct.unpack_from(">H", data, drop_start + 4)[0]
    drop_len = 4 + 2 + count * 2 + block_size
    crc_start = drop_start + drop_len
    if len(data) < crc_start + _CRC_LEN:
        return None
    got = struct.unpack_from(">H", data, crc_start)[0]
    want = crc16(data[len(SYNC) : crc_start])
    if got != want:
        return None
    return {
        "k_blocks": k_blocks,
        "block_size": block_size,
        "original_length": original_length,
        "droplet_bytes": data[drop_start : drop_start + drop_len],
    }


def decode(whitened_frames: List[str]) -> Optional[bytes]:
    """Decode a (possibly lossy, out-of-order) set of frames to the payload."""
    decoder = None
    original_length = None
    for wf in whitened_frames:
        f = decode_frame(wf)
        if f is None:
            continue  # dropped/corrupt frame — fountain tolerates it
        if decoder is None:
            decoder = FountainDecoder(f["k_blocks"], f["block_size"], f["original_length"])
            original_length = f["original_length"]
        decoder.add_droplet(unpack_droplet(f["droplet_bytes"], f["block_size"]))
        if decoder.is_complete():
            break
    if decoder is None or not decoder.is_complete():
        return None
    return decoder.get_data(original_length)
